### DISCUSSION TITLE
Fully Resolve Dependencies for OpenAPI Schema Generation (Authorization example)

### DISCUSSION THREAD

**Discussion by four43** [fastapi/fastapi#6323] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, Request
from fastapi.security import OAuth2PasswordBearer
from pydantic import BaseSettings


class Settings(BaseSettings):
    token_url: str

def get_settings() -> Settings:
    return Settings()


app = FastAPI()


def get_oauth2_scheme(settings=Depends(get_settings)):
    return OAuth2PasswordBearer(tokenUrl=settings.token_url)

async def use_token(request: Request, oauth2_scheme=Depends(get_oauth2_scheme)):
    return await oauth2_scheme(request)

@app.get("/items/")
async def read_items(token: str = Depends(use_token)):
    return {"token": token}
```

### Description

1. Run saved script, for example: `TOKEN_URL="token" uvicorn tmp:app`
2. Navigate to http://127.0.0.1:8000/docs
3. Notice there is no "Authorize" button in docs
4. Endpoint is protected, a request fails with a 401 error

### Wanted Solution

The Dependency Injection system doesn't seem to be fully resolving these dependencies so it can auto-detect the authentication mechanism. It seems to only resolve first surface level dependencies. This seems to apply to any Depends()
objects for authorization, either in route definitions or route signatures.

### Wanted Code

```python
(The example/reproduction code working would be great, no extra syntax needed)
```

### Operating System

Linux

### FastAPI Version

0.37.0

### Python Version

3.9

### Additional Context

Discussion in Discord here: https://discord.com/channels/772607987558252564/772631203689922562/942107142859673732

https://cdn.discordapp.com/attachments/772631203689922562/942107232315777094/unknown.png

**Comment by four43** [comment]
Oh man I am deep in this debugger here seeing the dependencies tried to be resolved at:

https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/utils.py#L321

By the stack trace it looks like it's tryin real hard! 
![image](https://user-images.githubusercontent.com/517006/153731189-706fce02-1e52-441b-946f-c812442bac17.png)

It seems to never get fully resolved and set as a `security_requirements` https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/utils.py#L201

Okay uncovering more now:
https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/utils.py#L147 
That seems to be checking for dependency, but they typically seem to be functions, so they're not resolved, they're just keeping their unresolved dependent-ness? In my example, it's `<function use_token at 0x7ffbbf62c160>` which returns security things, but it doesn't check for that.

**Comment by four43** [comment]
This is related to https://github.com/tiangolo/fastapi/issues/3317

**Comment by four43** [comment, marked as answer]
Since this is a bit of a duplicate, I'll close this and share my strategy:

I ended up subclassing the `SecurityBase` of my choosing. In my case it was `APIKeyQuery` and I wanted access to a Redis cache. Instead of configuring the authorizor via it's construction, `__init__()`, I ended up using `Depends()` in the `__call__()` method. Leveraging `@lru_cache` will hopefully keep it speedy enough. Thanks to BerteL on Discord for the help.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue 👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
